src/firecracker/examples/uffd/uffd_utils.rs

```diff
@@ -292,3 +292,92 @@ fn create_mem_regions(mappings: &Vec<GuestRegionUffdMapping>) -> Vec<MemRegion>
+
+        let mut uninit_runtime = Box::new(MaybeUninit::<Runtime>::uninit());
+        // We will use this pointer to bypass a bunch of Rust Safety
+        // for the sake of convenience.
+        let runtime_ptr = uninit_runtime.as_ptr() as *const Runtime;
+
+        let runtime_thread = std::thread::spawn(move || {
+            let tmp_file = TempFile::new().unwrap();
+            tmp_file.as_file().set_len(0x1000).unwrap();
+            let dummy_mem_path = tmp_file.as_path();
+
+            let file = File::open(dummy_mem_path).expect("Cannot open memfile");
+            let listener =
+                UnixListener::bind(dummy_socket_path).expect("Cannot bind to socket path");
+            let (stream, _) = listener.accept().expect("Cannot listen on UDS socket");
+            // Update runtime with actual runtime
+            let runtime = uninit_runtime.write(Runtime::new(stream, file));
+            runtime.run(|_: &mut UffdHandler| {});
+        });
+
+        // wait for runtime thread to initialize itself
+        std::thread::sleep(std::time::Duration::from_millis(100));
+
+        let stream =
+            UnixStream::connect(dummy_socket_path_clone).expect("Cannot connect to the socket");
+
+        let dummy_memory_region = vec![GuestRegionUffdMapping {
+            base_host_virt_addr: 0,
+            size: 0x1000,
+            offset: 0,
+        }];
+        let dummy_memory_region_json = serde_json::to_string(&dummy_memory_region).unwrap();
+
+        let dummy_file_1 = TempFile::new().unwrap();
+        let dummy_fd_1 = dummy_file_1.as_file().as_raw_fd();
+        stream
+            .send_with_fd(dummy_memory_region_json.as_bytes(), dummy_fd_1)
+            .unwrap();
+        // wait for the runtime thread to process message
+        std::thread::sleep(std::time::Duration::from_millis(100));
+        unsafe {
+            assert_eq!((*runtime_ptr).uffds.len(), 1);
+        }
+
+        let dummy_file_2 = TempFile::new().unwrap();
+        let dummy_fd_2 = dummy_file_2.as_file().as_raw_fd();
+        stream
+            .send_with_fd(dummy_memory_region_json.as_bytes(), dummy_fd_2)
+            .unwrap();
+        // wait for the runtime thread to process message
+        std::thread::sleep(std::time::Duration::from_millis(100));
+        unsafe {
+            assert_eq!((*runtime_ptr).uffds.len(), 2);
+        }
+
+        // there is no way to properly stop runtime, so
+        // we send a message with an incorrect memory region
+        // to cause runtime thread to panic
+        let error_memory_region = vec![GuestRegionUffdMapping {
+            base_host_virt_addr: 0,
+            size: 0,
+            offset: 0,
+        }];
+        let error_memory_region_json = serde_json::to_string(&error_memory_region).unwrap();
+        stream
+            .send_with_fd(error_memory_region_json.as_bytes(), dummy_fd_2)
+            .unwrap();
+
+        runtime_thread.join().unwrap_err();
+    }
```

Comment:
The issue is that test will stop at `let (stream, _) = listener.accept().expect("Cannot listen on UDS socket");` line which will wait forever, because the connection is only established later:  `let stream =
            UnixStream::connect(dummy_socket_path_clone).expect("Cannot connect to the socket");`

Reply:
```rust
#[test]
fn test_runtime() {
    let tmp_dir = TempDir::new().unwrap();
    let dummy_socket_path = tmp_dir.as_path().join("dummy_socket");
    let dummy_socket_path_clone = dummy_socket_path.clone();

    let pair = Arc::new(Mutex::new(OnceLock::new()), Condvar::new());
    let pair2 = pair.clone();

    // Open file for socket listener
    let tmp_file = TempFile::new().unwrap();
    tmp_file.as_file().set_len(0x1000).unwrap();
    let dummy_mem_path = tmp_file.as_path();
    let file = File::open(dummy_mem_path).unwrap();
    let listener = UnixListener::bind(dummy_socket_path).unwrap();
    
    // Start runtime thread
    let runtime_thread = std::thread::spawn(move || {
        let (lock, cvar) = &*pair2;
        // Accept connection
        let (stream, _) = listener.accept().unwrap();
        // Initalize runtime
        lock.lock().unwrap().set(Runtime::new(stream, file)).unwrap();
        // Notify test thread
        cvar.notify_one();
        // Run runtime
        lock.lock().unwrap().run(|_: &mut UffdHandler| {});
    });
    let (lock, cvar) = &*pair;

    // COnnect to socket
    let stream = UnixStream::connect(dummy_socket_path_clone).unwrap();

    let dummy_memory_region = vec![GuestRegionUffdMapping {
        base_host_virt_addr: 0,
        size: 0x1000,
        offset: 0,
    }];
    let dummy_memory_region_json = serde_json::to_string(&dummy_memory_region).unwrap();

    let dummy_file_1 = TempFile::new().unwrap();
    let dummy_fd_1 = dummy_file_1.as_file().as_raw_fd();
    stream
        .send_with_fd(dummy_memory_region_json.as_bytes(), dummy_fd_1)
        .unwrap();

    // Wait for runtime to be initialized.
    let _guard = cvar.wait_while(lock.lock().unwrap(), |pending| { pending.get().is_none() }).unwrap():

    // wait for the runtime thread to process message
    std::thread::sleep(std::time::Duration::from_millis(100));

    assert_eq!(lock.lock().unwrap().get().unwrap().uffds.len(), 1);

    let dummy_file_2 = TempFile::new().unwrap();
    let dummy_fd_2 = dummy_file_2.as_file().as_raw_fd();
    stream
        .send_with_fd(dummy_memory_region_json.as_bytes(), dummy_fd_2)
        .unwrap();
    // wait for the runtime thread to process message
    std::thread::sleep(std::time::Duration::from_millis(100));
    assert_eq!(lock.lock().unwrap().get().unwrap().uffds.len(), 2);

    // there is no way to properly stop runtime, so
    // we send a message with an incorrect memory region
    // to cause runtime thread to panic
    let error_memory_region = vec![GuestRegionUffdMapping {
        base_host_virt_addr: 0,
        size: 0,
        offset: 0,
    }];
    let error_memory_region_json = serde_json::to_string(&error_memory_region).unwrap();
    stream
        .send_with_fd(error_memory_region_json.as_bytes(), dummy_fd_2)
        .unwrap();

    runtime_thread.join().unwrap_err();
}
```

Reply:
I think you can open a PR with this change later, after your changes compile and run.

Reply:
Resolving this appears much more complex than I initially thought, leaving the test with an unlikely race condition is okay for now.